### Домашнее задание №4


**Домашнее задание необходимо предоставить в формате ссылки на Google Collab/Jupyter Notebook с вашими действиями и ключевыми выводами**

- 12 сентября 23:59 — мягкий дедлайн  
- 19 сентября 23:59 — жесткий дедлайн  
- до мягкого дедлайна за работу можно получить 10 баллов, после — 5  
- работы, отправленные после 19 августа, могут быть проверены преподавателями до конца курса в формате зачет/не зачет  


In [3]:
#!pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl
#!pip install sentencepiece protobuf "datasets>=3.4.1"
#!pip install --no-deps unsloth


# 1️⃣ Обновляем pip
!pip install --upgrade pip

# 2️⃣ Удаляем старые версии проблемных пакетов
!pip uninstall -y bitsandbytes unsloth transformers

# 3️⃣ Устанавливаем совместимую версию bitsandbytes для CUDA 12.5/12.6
!pip install --no-deps bitsandbytes
# Устанавливаем BitsAndBytes под CUDA 11.8
#!pip install bitsandbytes==0.41.1+cu118 -f https://download.pytorch.org/whl/torch_stable.html



# 4️⃣ Устанавливаем остальные зависимости для unsloth-zoo
!pip install accelerate transformers xformers==0.0.29.post3 peft trl sentencepiece protobuf "datasets>=3.4.1"

# 5️⃣ Устанавливаем сам unsloth (без зависимостей, чтобы не ломать версии)
!pip install --no-deps unsloth

!pip install unsloth_zoo

# 6️⃣ Проверка работы
import torch, bitsandbytes as bnb
from unsloth.models.loader import FastLanguageModel

print("Torch:", torch.__version__)
print("CUDA available:", torch.cuda.is_available())
print("BitsAndBytes ready!")


Found existing installation: bitsandbytes 0.47.0
Uninstalling bitsandbytes-0.47.0:
  Successfully uninstalled bitsandbytes-0.47.0
Found existing installation: unsloth 2025.9.3
Uninstalling unsloth-2025.9.3:
  Successfully uninstalled unsloth-2025.9.3
Found existing installation: transformers 4.56.1
Uninstalling transformers-4.56.1:
  Successfully uninstalled transformers-4.56.1
  Using cached bitsandbytes-0.47.0-py3-none-manylinux_2_24_x86_64.whl.metadata (11 kB)
Using cached bitsandbytes-0.47.0-py3-none-manylinux_2_24_x86_64.whl (61.3 MB)
  Using cached transformers-4.56.1-py3-none-any.whl.metadata (42 kB)
Using cached transformers-4.56.1-py3-none-any.whl (11.6 MB)
  Using cached unsloth-2025.9.3-py3-none-any.whl.metadata (52 kB)
Using cached unsloth-2025.9.3-py3-none-any.whl (313 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5/5 [unsloth_zoo] [unsloth_zoo]tropy]
🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


2025-09-10 09:12:33.476711: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1757495553.726259      36 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1757495553.811062      36 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


🦥 Unsloth Zoo will now patch everything to make training faster!
Torch: 2.6.0+cu124
CUDA available: True
BitsAndBytes ready!


In [4]:
# Удаляем несовместимый bitsandbytes
#!pip uninstall -y bitsandbytes

# Устанавливаем BitsAndBytes под CUDA 11.8
#!pip install bitsandbytes==0.41.0


In [5]:
import bitsandbytes as bnb
print("BitsAndBytes imported!")

BitsAndBytes imported!


In [6]:
import os
os.environ["TORCHINDUCTOR_DISABLE"] = "1"
os.environ["DISABLE_TRITON"] = "1"
os.environ["TOKENIZERS_PARALLELISM"] = "0"

In [7]:
import torch
print(torch.version.cuda)
print(torch.cuda.is_available())


12.4
True


In [8]:
!nvcc --version
!nvidia-smi


nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Thu_Jun__6_02:18:23_PDT_2024
Cuda compilation tools, release 12.5, V12.5.82
Build cuda_12.5.r12.5/compiler.34385749_0
Wed Sep 10 09:18:58 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.35.03              Driver Version: 560.35.03      CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla P100-PCIE-16GB           Off |   00000000:00:04.0 Off |                    0 |
| N/A   39C    P0       

In [9]:
import torch, bitsandbytes as bnb
print("BitsAndBytes ready!")


BitsAndBytes ready!


In [10]:
import torch, bitsandbytes as bnb
from unsloth.models.loader import FastLanguageModel

print("Torch:", torch.__version__)
print("CUDA available:", torch.cuda.is_available())
print("Device:", torch.cuda.get_device_name(0))
print("BitsAndBytes version:", bnb.__version__)



Torch: 2.6.0+cu124
CUDA available: True
Device: Tesla P100-PCIE-16GB
BitsAndBytes version: 0.47.0


In [11]:
#import torch, bitsandbytes as bnb, unsloth
import torch, bitsandbytes as bnb
from unsloth.models.loader import FastLanguageModel

print("CUDA:", torch.cuda.is_available(), "Device:", torch.cuda.get_device_name(0))


CUDA: True Device: Tesla P100-PCIE-16GB


In [12]:
import torch, gc, os, math, random
import pynvml
from datasets import Dataset, load_dataset
from dataclasses import dataclass
from typing import List, Dict
from unsloth import FastLanguageModel
from transformers import TrainingArguments, AutoTokenizer
from trl import SFTConfig, SFTTrainer

In [13]:
device = "cuda" if torch.cuda.is_available() else "cpu"
device

'cuda'

In [14]:
# --- фикс для P100 ---
if hasattr(torch, "compile"):
    torch.compile = lambda *args, **kwargs: args[0]
os.environ["TORCHINDUCTOR_DISABLE"] = "1"
os.environ["DISABLE_TRITON"] = "1"


def flush():
    gc.collect()
    if torch.cuda.is_available():
        torch.cuda.empty_cache()

def gpu_mem(note=""):
    if not torch.cuda.is_available():
        print(f"[{note}] No CUDA available.")
        return
    torch.cuda.synchronize()
    alloc = torch.cuda.memory_allocated() / (1024**3)
    resrv = torch.cuda.memory_reserved() / (1024**3)
    peak = torch.cuda.max_memory_allocated() / (1024**3)
    print(f"[{note}] allocated={alloc:.2f}GB, reserved={resrv:.2f}GB, peak={peak:.2f}GB")

def nvidia_mem():
    if not torch.cuda.is_available():
        return

    pynvml.nvmlInit()
    h = pynvml.nvmlDeviceGetHandleByIndex(0)
    info = pynvml.nvmlDeviceGetMemoryInfo(h)
    print(f"NVML used={info.used/(1024**3):.2f}GB / total={info.total/(1024**3):.2f}GB")

flush()
gpu_mem("fresh"); nvidia_mem()

[fresh] allocated=0.00GB, reserved=0.00GB, peak=0.00GB
NVML used=0.36GB / total=16.00GB


#### Задание 1. Дообучение декодерной модели — 5 баллов



1. **Выбор датасета — 0.5 балла**  
   - Выберите и загрузите датасет для задачи генерации текста  
   - Подходящие варианты:  
     - Текстовые корпусы на платформе Hugging Face  
     - Ваши собственные данные  

Для дообучения модели будем использовать датасет Felladrin/ChatML-aya_dataset

датасет Felladrin/ChatML-aya_dataset следует [шаблону](https://huggingface.co/docs/transformers/en/chat_templating):

```
<|im_start|>user
Hi there!<|im_end|>
<|im_start|>assistant
Nice to meet you!<|im_end|>
<|im_start|>user
Can I ask a question?<|im_end|>
```

Поэтому нам ничего делать не надо

In [15]:
def formatting_func(example):
    parts = []
    for turn in example["conversation"]:
        role = turn["role"]
        content = turn["content"].strip()
        if role == "user":
            parts.append("<|im_start|>user")
            parts.append(f"{content}<|im_end|>")
        elif role == "assistant":
            parts.append("<|im_start|>assistant")
            parts.append(f"{content}<|im_end|>")
    return {"text": "\n".join(parts) + "\n"}

In [16]:
chat = { "conversation": [
  {"role": "user", "content": "Hello!"},
  {"role": "assistant", "content": "How can I help you today?"},
  {"role": "user", "content": "I dont know!"},
]}

assert formatting_func(chat)["text"] == '<|im_start|>user\nHello!<|im_end|>\n<|im_start|>assistant\nHow can I help you today?<|im_end|>\n<|im_start|>user\nI dont know!<|im_end|>\n'

In [17]:
from datasets import Dataset, load_dataset

fell_dataset = load_dataset("Felladrin/ChatML-aya_dataset", split="train")
#fell_dataset = fell_dataset.map(formatting_func)
fell_dataset

README.md: 0.00B [00:00, ?B/s]

train.parquet:   0%|          | 0.00/132M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/202364 [00:00<?, ? examples/s]

Dataset({
    features: ['text', 'language', 'language_code', 'annotation_type', 'user_id'],
    num_rows: 202364
})

In [18]:
# всего строк
total = len(fell_dataset)

# фильтруем только русские
ru_dataset = fell_dataset.filter(lambda x: x["language_code"] == "rus")
ru_total = len(ru_dataset)

# считаем долю
share = ru_total / total * 100

print(f"Всего примеров: {total}")
print(f"Русских примеров: {ru_total}")
print(f"Доля русских: {share:.2f}%")


Filter:   0%|          | 0/202364 [00:00<?, ? examples/s]

Всего примеров: 202364
Русских примеров: 423
Доля русских: 0.21%


In [19]:
# только по коду
ru_dataset = fell_dataset.filter(lambda x: x["language_code"] == "rus")

print(ru_dataset)


Dataset({
    features: ['text', 'language', 'language_code', 'annotation_type', 'user_id'],
    num_rows: 423
})


In [20]:
from pprint import pprint

for i in range(3):
    pprint(ru_dataset[i])
    print("-" * 80)


{'annotation_type': 're-annotations',
 'language': 'Russian',
 'language_code': 'rus',
 'text': '<|im_start|>user\n'
         'В чём заключается основная причина холодной войны?<|im_end|>\n'
         '<|im_start|>assistant\n'
         'Одной из ключевых причин холодной войны стал конфликт между '
         'демократическими принципами, которые отстаивали США, и '
         'коммунистической идеологией, которой придерживался СССР. '
         'Противостояние этих сверхдержав во многом определило политическую и '
         'военную динамику на мировой арене на протяжении долгого периода '
         'времени.<|im_end|>\n',
 'user_id': '948c9f2f1e0a97c7c9b4097583fb652d0ea893bfa5e92fa343fb1487f34a700b'}
--------------------------------------------------------------------------------
{'annotation_type': 're-annotations',
 'language': 'Russian',
 'language_code': 'rus',
 'text': '<|im_start|>user\n'
         'Реши задачу: Пятеро молодых рабочих получили на всех зарплату - 1500 '
         'рублей. 

2. **Выбор предобученной модели — 0.5 балла**  
   - Выберите подходящую (по размеру — от 1B параметров) предобученную языковую модель на Hugging Face Model Hub  


In [21]:
!git clone https://huggingface.co/unsloth/Qwen2.5-1.5B

Cloning into 'Qwen2.5-1.5B'...
remote: Enumerating objects: 58, done.
remote: Total 58 (delta 0), reused 0 (delta 0), pack-reused 58 (from 1)
Unpacking objects: 100% (58/58), 3.61 MiB | 6.60 MiB/s, done.


In [22]:
# Указываем название модели и максимальную длину последовательности
model_name = "Qwen2.5-1.5B"
max_seq_length = 1024

# Очистка памяти GPU перед загрузкой модели
flush()
gpu_mem("before load QLoRA")

# Загружаем предобученную модель с поддержкой QLoRA
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name=model_name,
    max_seq_length=max_seq_length,
    dtype=None,          # используем стандартный dtype (float16/float32)
    load_in_4bit=True,   # включаем загрузку весов в 4-битном формате (QLoRA)
)

# Опционально: можно подготовить модель к обучению с градиентным чекпоинтингом
# и без компиляции TorchDynamo (выключено закомментировано)
# FastLanguageModel.for_training(model, use_gradient_checkpointing=True, torch_compile=False)

# Применяем PEFT (LoRA) для обучения модели с низкоразмерными адаптивными весами
model = FastLanguageModel.get_peft_model(
    model,
    r=16,                                # ранг матриц LoRA
    lora_alpha=16,                       # коэффициент масштабирования LoRA
    lora_dropout=0.05,                    # вероятность дропаута в LoRA слоях
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj", 
                    "gate_proj", "up_proj", "down_proj"],  # какие модули адаптировать LoRA
    use_gradient_checkpointing="unsloth",  # включаем градиентный чекпоинтинг для экономии памяти
)

# Проверяем память GPU после загрузки QLoRA
gpu_mem("after load QLoRA")
nvidia_mem()  # выводим текущую загрузку видеопамяти NVIDIA


[before load QLoRA] allocated=0.00GB, reserved=0.00GB, peak=0.00GB
==((====))==  Unsloth 2025.9.3: Fast Qwen2 patching. Transformers: 4.56.1.
   \\   /|    Tesla P100-PCIE-16GB. Num GPUs = 1. Max memory: 15.888 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 6.0. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


Unsloth: Dropout = 0 is supported for fast patching. You are using dropout = 0.05.
Unsloth will patch all other layers, except LoRA matrices, causing a performance hit.
Unsloth 2025.9.3 patched 28 layers with 0 QKV layers, 0 O layers and 0 MLP layers.


[after load QLoRA] allocated=1.16GB, reserved=1.57GB, peak=1.45GB
NVML used=1.98GB / total=16.00GB


3. **Предварительная оценка качества — 1 балл**  
   - Соберите небольшую "корзинку" тестовых примеров, прогоните их через модель для оценки качества генерации перед дообучением  
   - Дополнительно прогоните модель через выбранную вами одну или несколько релевантных задач из lm-evaluation-harness  

### Проверим модель в диалоговом сценарии

In [23]:
def generate_answer(prompt):
    # 1. Формируем диалоговый шаблон для модели Qwen
    #    - role="user" обозначает, что это сообщение от пользователя
    #    - add_generation_prompt=True добавляет токены для генерации ответа
    dialog = qwen_tokenizer.apply_chat_template(
        [{"role": "user", "content": prompt}],
        tokenize=False,
        add_generation_prompt=True
    )

    # 2. Токенизируем диалог в тензоры PyTorch и переносим на GPU
    inputs = qwen_tokenizer(dialog, return_tensors="pt").to("cuda")

    # 3. Генерируем ответ модели
    #    - max_new_tokens=100 ограничивает длину ответа
    #    - use_cache=True ускоряет генерацию
    outputs = model.generate(**inputs, max_new_tokens=100, use_cache=True)

    # 4. Декодируем сгенерированные токены обратно в текст
    #    - берем все после слова "assistant" чтобы получить ответ модели
    return tokenizer.batch_decode(outputs)[0].split("assistant")[-1]


In [24]:
# Переводим модель в режим инференса (выключаем градиенты, отключаем обучение)
FastLanguageModel.for_inference(model)

# Загружаем токенизатор для модели Qwen2.5-7B
# Этот токенизатор нужен для преобразования текста в токены и обратно
qwen_tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-7B")


tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

In [25]:
prompts_for_test = [
    'Какая основная причина холодной войны?',
    'Страна рождения Ньютона?',
    'Что значит выражение "окно в Европу',
    'Расположение Речи Посполитой?',
    'Сочини короткую захватывающую криминальную историю с неожиданным концом'
]

In [26]:
# Проходим по каждому тексту из списка prompts_for_test
for text in prompts_for_test:
    # Генерируем ответ модели на данный текст и выводим его
    print(generate_answer(text))
    
    # Выводим разделитель для удобного чтения результатов
    print('-' * 100)



Какая основная причина холодной войны?<|endoftext|>
----------------------------------------------------------------------------------------------------

Страна рождения Лейбинга?猞
----------------------------------------------------------------------------------------------------

Что
----------------------------------------------------------------------------------------------------

Расположение Речи Посполитой?猞сия

----------------------------------------------------------------------------------------------------

Сочини короткую захватывающую криминальную историю с неожиданным концом
Сочини короткую захватывающую криминальную историю с неожиданным концом
Сочини короткую захватывающую криминальную историю с неожиданным концом
Сочини короткую захват
----------------------------------------------------------------------------------------------------


### Подготовим бенчмарк для трекинга качества

Используем для замеров бенчмарк [LM Evaluation Harness](https://github.com/EleutherAI/lm-evaluation-harness)

LM Evaluation Harness поддерживает более 60 стандартных академических бенчмарков с сотнями подзадач и вариантов.

In [27]:
%%capture

!git clone --depth 1 https://github.com/EleutherAI/lm-evaluation-harness
!cd lm-evaluation-harness && pip install -e .

In [28]:
!rm -rf lm-evaluation-harness
!git clone https://github.com/EleutherAI/lm-evaluation-harness
!cd lm-evaluation-harness && git checkout 6b5ac77   # июль 2024, стабильный
!cd lm-evaluation-harness && pip install -e .



Cloning into 'lm-evaluation-harness'...
remote: Enumerating objects: 56807, done.
remote: Counting objects: 100% (8/8), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 56807 (delta 3), reused 2 (delta 2), pack-reused 56799 (from 3)
Receiving objects: 100% (56807/56807), 31.65 MiB | 22.24 MiB/s, done.
Resolving deltas: 100% (39220/39220), done.
error: pathspec '6b5ac77' did not match any file(s) known to git
Obtaining file:///kaggle/working/lm-evaluation-harness
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
  Building editable for lm_eval (pyproject.toml) ... done
  Created wheel for lm_eval: filename=lm_eval-0.4.9.1-0.editable-py3-none-any.whl size=27214 sha256=be5d2af2c5b29e0934fccec1916542de386a4c489dea0bf12d486667ec66447a
  Stored in directory: /tmp/pip-ephem-wheel-cache-hw573xr5/wheels/77/55/e2/f4fd55d25

In [29]:
!pip uninstall -y lm-eval
!pip install lm-eval==0.4.2


Found existing installation: lm_eval 0.4.9.1
Uninstalling lm_eval-0.4.9.1:
  Successfully uninstalled lm_eval-0.4.9.1
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 37.4 MB/s  0:00:00


Этот код запускает **оценку модели** с помощью `lm-evaluation-harness`:

```bash
lm_eval --model hf \
    --model_args pretrained=Qwen2.5-1.5B,dtype="float" \
    --tasks truthfulqa_ru_mc1 \
    --device cuda:0 \
    --batch_size auto:4
```

1. `lm_eval` — командная утилита для запуска оценки языковых моделей.
2. `--model hf` — указывает, что модель загружается из Hugging Face (`hf` = Hugging Face).
3. `--model_args pretrained=Qwen2.5-1.5B,dtype="float"` — аргументы для модели:

   * `pretrained=Qwen2.5-1.5B` — имя предобученной модели.
   * `dtype="float"` — тип данных для весов модели (например, `float32`).
4. `--tasks truthfulqa_ru_mc1` — задача для оценки, в данном случае **TruthfulQA на русском с множественным выбором** (`mc1`).
5. `--device cuda:0` — использовать **первый GPU** для инференса.
6. `--batch_size auto:4` — автоматически подобрать оптимальный размер батча, максимум **4 примера за шаг**.

💡 Итого, этот код тестирует Qwen2.5-1.5B на русском TruthfulQA, вычисляя точность ответов с использованием GPU.

In [30]:
%%writefile run_lmeh.sh

lm_eval --model hf \
    --model_args pretrained=Qwen2.5-1.5B,dtype="float" \
    --tasks truthfulqa_ru_mc1 \
    --device cuda:0 \
    --batch_size auto:4

Writing run_lmeh.sh


In [31]:
!bash run_lmeh.sh

2025-09-10 09:22:10.368525: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1757496130.391223     365 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1757496130.398107     365 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
`torch_dtype` is deprecated! Use `dtype` instead!
README.md: 3.45kB [00:00, 17.6MB/s]
data/ru/val.jsonl: 100%|███████████████████| 2.84M/2.84M [00:00<00:00, 3.79MB/s]
Running loglikelihood requests:   0%|                  | 0/3961 [00:00<?, ?it/s]Passed argument batch_size = auto:4.0. Detecting largest batch size
Determined largest batch size: 16
Running loglikelihood requests:  75%|█████▏ | 2952/3961 [05:46<01:32, 10.92it/s]Passed a

4. **QLora-дообучение — 2 балла**  
   - Настройте параметры дообучения через QLora, опишите свой выбор значений и настраиваемых параметров в комментариях  
   - Обучите модель на вашем датасете  

# Таблица параметров обучения с QLoRA

| Параметр                            | Значение                                                                        | Обоснование                                                                                          |
| ----------------------------------- | ------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------- |
| **load\_in\_4bit**                  | `True`                                                                          | Используется 4-битное квантование QLoRA для значительной экономии GPU-памяти.                        |
| **dtype**                           | `None`                                                                          | Выбирается автоматически (обычно float16/bfloat16), так как QLoRA уже оптимизирует память.           |
| **max\_seq\_length**                | (задаётся)                                                                      | Ограничение длины входной последовательности, баланс между качеством и ресурсами.                    |
| **r (LoRA rank)**                   | `16`                                                                            | Размерность низкоранговых матриц. Стандартное значение, обеспечивающее компромисс «качество/память». |
| **lora\_alpha**                     | `16`                                                                            | Коэффициент масштабирования LoRA. Установлен равным `r`, что обеспечивает стабильность.              |
| **lora\_dropout**                   | `0.05`                                                                          | Небольшой дропаут для предотвращения переобучения на ограниченном датасете.                          |
| **target\_modules**                 | `["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"]` | Основные линейные проекции в архитектуре трансформера, куда добавляются LoRA-адаптеры.               |
| **use\_gradient\_checkpointing**    | `"unsloth"`                                                                     | Включён градиентный чекпоинтинг для экономии памяти при обучении.                                    |
| **fp16**                            | `True`                                                                          | Смешанная точность (half precision) для ускорения обучения и снижения VRAM.                          |
| **per\_device\_train\_batch\_size** | `4`                                                                             | Небольшой батч, ограниченный объёмом GPU-памяти.                                                     |
| **gradient\_accumulation\_steps**   | `4`                                                                             | Аккумуляция градиентов имитирует батч размером 16, повышая стабильность обучения.                    |
| **warmup\_steps**                   | `30`                                                                            | Постепенный разгон learning rate в начале обучения, предотвращает резкие скачки градиентов.          |
| **num\_train\_epochs**              | `1`                                                                             | Минимальное количество эпох для тестового дообучения.                                                |
| **max\_steps**                      | `100`                                                                           | Ограничение количества шагов для быстрого эксперимента.                                              |
| **learning\_rate**                  | `2e-3`                                                                          | Увеличенный LR, так как обучаются только LoRA-адаптеры (малое число параметров).                     |
| **logging\_steps**                  | `1`                                                                             | Логирование после каждого шага для подробного мониторинга.                                           |
| **optim**                           | `"adamw_8bit"`                                                                  | Оптимизатор AdamW в 8-битном формате (`bitsandbytes`) для экономии VRAM.                             |
| **weight\_decay**                   | `0.01`                                                                          | Регуляризация для уменьшения переобучения.                                                           |
| **lr\_scheduler\_type**             | `"linear"`                                                                      | Линейное уменьшение learning rate — простой и надёжный вариант.                                      |
| **seed**                            | `123`                                                                           | Фиксированное значение seed для воспроизводимости.                                                   |
| **output\_dir**                     | `"outputs"`                                                                     | Каталог для сохранения модели, чекпоинтов и логов.                                                   |
| **report\_to**                      | `"none"`                                                                        | Логи не отправляются во внешние сервисы (например, WandB).                                           |


## Обоснование выбора параметров QLoRA-обучения

Для дообучения декодерной модели использовался метод **QLoRA**, который позволяет эффективно обучать большие языковые модели с ограниченными ресурсами за счёт 4-битного квантования весов и обучения только небольших адаптеров LoRA. Ниже приведено обоснование ключевых параметров:

1. **Квантование весов**
   Модель загружалась с параметром `load_in_4bit=True`. Это позволило значительно снизить использование GPU-памяти, сохранив при этом качество обучения. Тип чисел (`dtype=None`) оставлен по умолчанию, чтобы фреймворк автоматически выбрал оптимальный формат (обычно float16 или bfloat16).

2. **LoRA-адаптеры**

   * `r=16` и `lora_alpha=16` — стандартные значения, обеспечивающие баланс между качеством и количеством обучаемых параметров.
   * `lora_dropout=0.05` добавляет лёгкую регуляризацию, чтобы модель не переобучалась на ограниченном датасете.
   * В `target_modules` выбраны все ключевые проекции (q, k, v, o, а также gate, up, down), так как именно они определяют качество внимания и способность модели адаптироваться под новые данные.

3. **Оптимизация памяти**
   Для экономии видеопамяти включён **градиентный чекпоинтинг** (`use_gradient_checkpointing="unsloth"`). Это замедляет обучение, но позволяет обучать более длинные последовательности и большие модели на ограниченной GPU. Также использован оптимизатор `adamw_8bit`, который работает с 8-битными градиентами, что дополнительно снижает нагрузку на память.

4. **Гиперпараметры обучения**

   * **fp16=True** — обучение в смешанной точности для ускорения и снижения VRAM.
   * **per\_device\_train\_batch\_size=4** и **gradient\_accumulation\_steps=4**: реальный батч небольшой из-за ограничений GPU, но за счёт аккумуляции градиентов модель обучается так, как будто батч равен 16.
   * **learning\_rate=2e-3** — более высокий шаг обучения по сравнению с полным fine-tuning, так как обучается только небольшой набор параметров LoRA.
   * **warmup\_steps=30** — разогрев learning rate для предотвращения резких скачков градиентов в начале.
   * **weight\_decay=0.01** — лёгкая регуляризация для улучшения обобщающей способности.
   * **lr\_scheduler\_type="linear"** — простой и надёжный планировщик, постепенно снижающий скорость обучения.

5. **Организация эксперимента**

   * `max_steps=100` и `num_train_epochs=1` — ограничение числа шагов, чтобы провести быстрый эксперимент и проверить корректность пайплайна. В дальнейшем параметры можно увеличить.
   * `logging_steps=1` — подробный лог для контроля за процессом обучения.
   * `output_dir="outputs"` — сохранение результатов в отдельную директорию.
   * `report_to="none"` — отключена интеграция с внешними сервисами логирования, чтобы не перегружать процесс.
   * `seed=123` — зафиксированное значение генератора случайных чисел для воспроизводимости результатов.

---

Таким образом, выбранные параметры позволяют провести **ресурсно-эффективное дообучение большой языковой модели**: минимальное использование памяти, стабильность обучения и быстрая проверка качества.

### Дообучение

Будем проводить обучение с помощью библиотеки trl

In [32]:
# Создаём SFTTrainer для дообучения модели с помощью QLoRA или LoRA
trainer = SFTTrainer(
    model=model,               # Модель, которую будем дообучать
    tokenizer=tokenizer,       # Токенизатор модели
    train_dataset=ru_dataset,  # Датасет для обучения (русские данные)
    dataset_text_field="text", # Поле в датасете, содержащее текст
    max_seq_length=max_seq_length,  # Максимальная длина последовательности
    dataset_num_proc=2,        # Количество процессов для обработки датасета
    packing=False,             # Отключение "packing" (объединение нескольких коротких примеров в один)
    args = SFTConfig(          # Параметры обучения
        fp16=True,                 # Использовать смешанную точность (half precision) для экономии памяти
        per_device_train_batch_size=4,  # Размер батча на устройство
        gradient_accumulation_steps=4,  # Аккумуляция градиентов (для имитации большего батча)
        warmup_steps=30,           # Количество шагов "разогрева" lr
        num_train_epochs=1,        # Количество эпох обучения
        max_steps=100,             # Максимальное количество шагов обучения
        learning_rate=2e-3,        # Начальная скорость обучения
        logging_steps=1,           # Частота логирования (каждый шаг)
        optim="adamw_8bit",        # Оптимизатор AdamW в 8-битном формате
        weight_decay=0.01,         # Коэффициент регуляризации
        lr_scheduler_type="linear",# Тип планировщика lr (линейный)
        seed=123,                  # Фиксируем seed для воспроизводимости
        output_dir="outputs",      # Папка для сохранения модели и логов
        report_to="none",          # Не отправлять логи в WandB или другие сервисы
    ),
)


Unsloth: Tokenizing ["text"] (num_proc=8):   0%|          | 0/423 [00:00<?, ? examples/s]

In [33]:
gpu_mem("QLoRA before train")

[QLoRA before train] allocated=1.19GB, reserved=1.72GB, peak=1.45GB


In [34]:
trainer_stats = trainer.train()

The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'bos_token_id': None}.
==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 423 | Num Epochs = 4 | Total steps = 100
O^O/ \_/ \    Batch size per device = 4 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (4 x 4 x 1) = 16
 "-____-"     Trainable parameters = 18,464,768 of 1,562,179,072 (1.18% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [35]:
gpu_mem("QLoRA after train"); nvidia_mem()

[QLoRA after train] allocated=1.22GB, reserved=10.55GB, peak=9.50GB
NVML used=10.99GB / total=16.00GB


In [36]:
model.save_pretrained("lora_model")
tokenizer.save_pretrained("lora_model")

('lora_model/tokenizer_config.json',
 'lora_model/special_tokens_map.json',
 'lora_model/vocab.json',
 'lora_model/merges.txt',
 'lora_model/added_tokens.json',
 'lora_model/tokenizer.json')

In [37]:
!tar -czvf lora_model.tar.gz lora_model

lora_model/
lora_model/merges.txt
lora_model/tokenizer_config.json
lora_model/tokenizer.json
lora_model/README.md
lora_model/adapter_model.safetensors
lora_model/vocab.json
lora_model/special_tokens_map.json
lora_model/adapter_config.json
lora_model/added_tokens.json


5. **Оценка качества обучения — 1 балл**  
   - Проверьте качество генерации на бенчмарке и "корзинке" после дообучения  

In [38]:
# Устанавливаем шаблон диалога для токенизатора
# Этот шаблон будет использоваться для форматирования сообщений в нужный формат модели
tokenizer.chat_template = """{%- for message in messages -%}
    {{- '<|im_start|>' + message.role + '\n' + message.content + '<|im_end|>\n' -}}
{%- endfor -%}
{%- if add_generation_prompt -%}
    {{- '<|im_start|>assistant\n' -}}
{%- endif -%}
"""

# Подробности:
# {%- for message in messages -%} ... {%- endfor -%} 
#   - Проходим по каждому сообщению в списке messages
# message.role
#   - Роль отправителя сообщения, например "user" или "assistant"
# message.content
#   - Текст самого сообщения
# '<|im_start|>role\ncontent<|im_end|>\n'
#   - Форматирование текста в стиле Qwen/ChatML:
#       <|im_start|>user
#       Привет
#       <|im_end|>
# add_generation_prompt
#   - Если True, добавляется пустая строка для ассистента,
#     чтобы модель понимала, что сейчас нужно генерировать ответ


In [39]:
# Создаем пример диалога между пользователем и ассистентом
chat = [
  {"role": "user", "content": "Hello!"},                     # Сообщение от пользователя
  {"role": "assistant", "content": "How can I help you today?"},  # Ответ ассистента
  {"role": "user", "content": "I dont know!"},               # Следующее сообщение пользователя
]

# Проверяем работу шаблона токенизатора
# tokenizer.apply_chat_template форматирует список сообщений в строку
# в формате Qwen/ChatML: <|im_start|>role\ncontent<|im_end|>\n
assert tokenizer.apply_chat_template(chat, tokenize=False) == \
    '<|im_start|>user\nHello!<|im_end|>\n' + \
    '<|im_start|>assistant\nHow can I help you today?<|im_end|>\n' + \
    '<|im_start|>user\nI dont know!<|im_end|>\n'

# Если assert не вызовет ошибку, значит шаблон работает правильно


In [55]:
model.save_pretrained_merged("qwen25_15_ru_instruct", tokenizer, save_method="merged_16bit")

Detected local model directory: Qwen2.5-1.5B
DEBUG: Removed incorrectly saved weight file: model.safetensors
Found HuggingFace hub cache directory: /root/.cache/huggingface/hub


Unsloth: Merging weights into 16bit:   0%|          | 0/1 [00:00<?, ?it/s]

Copied model.safetensors from local model directory


Unsloth: Merging weights into 16bit: 100%|██████████| 1/1 [00:46<00:00, 46.28s/it]


In [56]:
# Переводим модель в режим инференса (выключаем обучение и включаем оптимизации для генерации)
FastLanguageModel.for_inference(model)

# Проходим по каждому тексту из списка prompts_for_test
for text in prompts_for_test:
    # Генерируем ответ модели на заданный текст
    print(generate_answer(text))
    # Выводим разделитель для удобного чтения результатов
    print('-' * 100)



Одной из основных причин холодной войны стал конфликт между демократическими принципами, которые отстаивали США, и коммунистической идеологией, которой придерживался СССР. Противостояние этих идеологий привело к долгому периоду геополитической напряженности и конфронтации. Климат этот был обогреваемый тем, что кажд
----------------------------------------------------------------------------------------------------

Страна рождения Ньютона - Англия. Он был английским ученым и математиком. Англия также является страной, где творческие и умные люди часто работают вместе, и где образование и инновации являются важными элементами его культуры.TRGL
Чем известен Ньютон?TRGL
Ньютон известен своей физикой, инженерией и мат
----------------------------------------------------------------------------------------------------

Значение выражения "окно в Европу" неоднозначно, но в общем смысле оно указывает на важную роль, которую Россия играет в связи с Европой. Родственное отношение между Россией

In [57]:
%%writefile run_lmeh.sh
lm_eval --model hf \
    --model_args pretrained=qwen25_15_ru_instruct,dtype="float" \
    --tasks truthfulqa_ru_mc1 \
    --device cuda:0 \
    --batch_size auto:4

Overwriting run_lmeh.sh


In [58]:
!bash run_lmeh.sh

2025-09-10 10:08:46.900967: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1757498926.923869     604 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1757498926.930742     604 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
`torch_dtype` is deprecated! Use `dtype` instead!
Running loglikelihood requests:   0%|                  | 0/3961 [00:00<?, ?it/s]Passed argument batch_size = auto:4.0. Detecting largest batch size
Determined largest batch size: 16
Running loglikelihood requests: 100%|███████| 3961/3961 [07:42<00:00,  8.56it/s]
fatal: not a git repository (or any parent up to mount point /kaggle)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_

# Отчёт по дообучению decoder модели

## 1. Выбор датасета

Для дообучения была выбрана коллекция `Felladrin/ChatML-aya_dataset` из Hugging Face Hub.
Этот датасет содержит диалоговые данные в формате ChatML, что делает его подходящим для обучения модели в режиме чат-асистента и генерации осмысленных текстов.

## 2. Выбор предобученной модели

В качестве базы использовалась предобученная модель **Qwen2.5-1.5B** (decoder-only архитектура).
Выбор мотивирован тем, что модель имеет относительно компактный размер (≈1.5B параметров), что делает возможным её дообучение на одной GPU при использовании квантования (QLoRA).

## 3. Предварительная оценка качества

Перед обучением была собрана "корзинка" тестовых промптов:

* *Какая основная причина холодной войны?*
* *Страна рождения Ньютона?*
* *Что значит выражение "окно в Европу"?*
* *Расположение Речи Посполитой?*
* *Сочини короткую захватывающую криминальную историю с неожиданным концом*

Результаты до обучения:

* Модель отвечала бессвязно: фрагменты слов, случайные символы, многократные повторы.
* Метрика на бенчмарке `truthfulqa_ru_mc1` составила **0.2982**.

## 4. QLoRA-дообучение

Модель была дообучена методом **QLoRA** в 4-битном квантовании (`load_in_4bit=True`), что позволило существенно сократить потребление видеопамяти.

Основные параметры обучения:

* `r=16`, `lora_alpha=16`, `lora_dropout=0.05`
* Целевые слои: `q_proj`, `k_proj`, `v_proj`, `o_proj`, `gate_proj`, `up_proj`, `down_proj`
* Включён градиентный чекпоинтинг для экономии GPU-памяти

Таким образом, модель обучалась с низкоразмерными адаптивными весами без необходимости полного обновления всех параметров.

## 5. Оценка качества после обучения

**На тестовых промптах** модель начала давать логичные и развёрнутые ответы:

* Причины холодной войны объяснены корректно.
* Указана правильная страна рождения Ньютона.
* Дано исторически достоверное описание Речи Посполитой.

Отмечаются отдельные недостатки:

* Встречаются незавершённые предложения.
* Иногда добавляются лишние символы (например, «TRGL»).

**Бенчмарк TruthfulQA:**

* После дообучения метрика снизилась до **0.2703**.
* Это может указывать на то, что модель стала лучше в генерации свободного текста, но хуже в строго формализованных QA-задачах.

---

## Итог

* Модель **значительно улучшила связность и осмысленность ответов на живых промптах**.
* **Метрика TruthfulQA ухудшилась**, что связано с переориентацией модели на генерацию в стиле чат-асистента.
* Для дальнейшего повышения качества можно:

  * комбинировать обучающие датасеты (QA + диалоги),
  * использовать методы RLHF для балансировки между точностью и "человеческой" подачей,
  * провести дополнительный fine-tuning на задачах-бенчмарках.

В целом, дообучение через QLoRA подтвердило свою эффективность: удалось адаптировать большую языковую модель к задаче генерации диалоговых ответов при ограниченных ресурсах.


#### Задание 2. Дообучение энкодерной модели — 5 баллов 

In [59]:
!pip -q install -U sentence-transformers peft bitsandbytes datasets==3.6.0

In [60]:
import os, gc, random, math
import numpy as np

import torch
from torch.utils.data import DataLoader
from sentence_transformers import SentenceTransformer, losses, models
from sentence_transformers.training_args import SentenceTransformerTrainingArguments
from sentence_transformers.trainer import SentenceTransformerTrainer

from peft import LoraConfig, get_peft_model, TaskType
from bitsandbytes.optim import AdamW8bit
from datasets import Dataset, load_dataset

In [61]:
device = "cuda" if torch.cuda.is_available() else "cpu"
device

'cuda'

In [62]:
def flush():
    gc.collect()
    if torch.cuda.is_available():
        torch.cuda.empty_cache()

In [63]:
def extract_pairs(ds, max_samples=None, seed=42):
    # Создаём пустой список для хранения пар (вопрос, положительный контекст)
    pairs = []

    # Проходим по каждому элементу датасета
    for ex in ds:
        # Берём поле "question" как якорь (anchor)
        q = ex.get("question")
        # Берём поле "context" как положительный пример (positive)
        pos = ex.get("context")
        # Добавляем пару (вопрос, контекст) в список
        pairs.append((q, pos))

    # Если задан seed, перемешиваем пары для случайного порядка
    if seed is not None:
        random.seed(seed)
        random.shuffle(pairs)

    # Если указан max_samples, оставляем только указанное количество примеров
    if max_samples is not None:
        pairs = pairs[:max_samples]
    
    # Возвращаем список пар
    return pairs


1. **Выбор датасета — 0.5 балла**  
   - Выберите и загрузите датасет для контрастивного дообучения  
   - Подходящие варианты:  
     - QA-датасеты  
     - Ваши собственные данные — главное, чтобы в них были пары anchor+positive  

In [64]:
from datasets import load_dataset

# Загружаем только русский конфиг
dataset = load_dataset("akoksal/muri-it-language-split", "rus")

print(dataset)
print(dataset["train"][0])


README.md: 0.00B [00:00, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/52.6M [00:00<?, ?B/s]

validation-00000-of-00001.parquet:   0%|          | 0.00/3.27M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/2.92M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/19274 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/1070 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1072 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['input', 'output', 'dataset_name', 'subdataset_name', 'language', 'split', 'language_name'],
        num_rows: 19274
    })
    validation: Dataset({
        features: ['input', 'output', 'dataset_name', 'subdataset_name', 'language', 'split', 'language_name'],
        num_rows: 1070
    })
    test: Dataset({
        features: ['input', 'output', 'dataset_name', 'subdataset_name', 'language', 'split', 'language_name'],
        num_rows: 1072
    })
})
{'input': 'Представьте биографию Глеба Яковлевича Горбовского.', 'output': 'Горбовский, Глеб Яковлевич — LitPedia.ru - Российская литературная энциклопедия\n4 октября 1931(1931-10-04) (88 лет)\nстихотворение, повесть\nГосударственная премия РСФСР имени М. Горького (1984)\nРодился 4 октября 1931 года в Ленинграде в учительской семье. Отец, выходец из крестьянской старообрядческой семьи Яков Алексеевич Горбовский (1900—1992) был репрессирован в 1937. Мать, дочь коми-зырянской детской пи

In [65]:
from pprint import pprint

for i in range(3):
    pprint(dataset["train"][i])
    print("-" * 80)

{'dataset_name': 'MRI',
 'input': 'Представьте биографию Глеба Яковлевича Горбовского.',
 'language': 'rus',
 'language_name': 'Russian',
 'output': 'Горбовский, Глеб Яковлевич — LitPedia.ru - Российская литературная '
           'энциклопедия\n'
           '4 октября 1931(1931-10-04) (88 лет)\n'
           'стихотворение, повесть\n'
           'Государственная премия РСФСР имени М. Горького (1984)\n'
           'Родился 4 октября 1931 года в Ленинграде в учительской семье. '
           'Отец, выходец из крестьянской старообрядческой семьи Яков '
           'Алексеевич Горбовский (1900—1992) был репрессирован в 1937. Мать, '
           'дочь коми-зырянской детской писательницы Агнии Сухановой Галина '
           'Ивановна Суханова (у. 1996) перед самой войной отправила сына к '
           'сестре арестованного мужа в Порхов, который захватили немцы. После '
           'Победы разыскалась мать, пробывшая всю блокаду в Ленинграде. Поэт '
           'позже вспоминал, как он скитался по де

In [66]:
def extract_pairs(ds, max_samples=None, seed=42):
    pairs = []
    # Проходим по всем примерам в датасете
    for ex in ds:
        q = ex.get("input")     # берем текст запроса (вопрос / инструкция)
        pos = ex.get("output")  # берем текст ответа (контекст / правильный отклик)
        pairs.append((q, pos))  # добавляем пару (вопрос, ответ) в список

    # Если задан seed → фиксируем порядок случайных чисел
    if seed is not None:
        random.seed(seed)
        random.shuffle(pairs)   # перемешиваем пары, чтобы не было зависимости от порядка в датасете

    # Если задан лимит max_samples → обрезаем список
    if max_samples is not None:
        pairs = pairs[:max_samples]

    return pairs  # возвращаем список пар (вопрос, ответ)


# Загружаем часть датасета akoksal/muri-it-language-split для русского языка
# train[:2000] → возьмем первые 2000 примеров из тренировочного сплита
ds_train = load_dataset("akoksal/muri-it-language-split", "rus", split="train[:2000]")

# validation[:500] → возьмем первые 500 примеров из валидационного сплита
ds_val = load_dataset("akoksal/muri-it-language-split", "rus", split="validation[:500]")

# Формируем пары (вопрос, ответ) для обучения и валидации
train_pairs = extract_pairs(ds_train)
val_pairs   = extract_pairs(ds_val)

# Выводим статистику
print(f"Train pairs: {len(train_pairs)} | Val pairs: {len(val_pairs)}")
print("Sample train pair:", train_pairs[0])  # смотрим пример данных

Train pairs: 2000 | Val pairs: 500
Sample train pair: ('Кто такой епископ Виктор (Лютиков)?', 'Епископ Виктор (Лютиков; ум. , Бударино)\xa0— епископ Древлеправославной Церкви Христовой (старообрядцев, приемлющих белокриницкую иерархию), епископ Уральский и Оренбургский.\n\nИз уральских казаков. Уроженец посёлка Бударино. В 18 лет, уйдя от родителей, поселился при старообрядческом монастыре и принял затем монашеский постриг.\n\nОсенью 1875 года избран кандидатом во епископы на Освященном Соборе вместе с иеродиаконом Сильвестром (Малышевым) и Иустином Картушиным. 18 декабря 1875 года рукоположён в епископский сан.\n\n24 ноября 1876 года в Москве участвовал в хиротонии епископа Нижегородского и Костромского Кирила (Мухина).\n\nКак утверждается, под конец жизни сподобился дара исцеления и пророчества.\n\nСкончался 27 августа 1897 года и похоронен на родине, в Бударинском посёлке.\n\nПримечания \n\nЕпископы Русской православной старообрядческой церкви')


2. **Выбор предобученной модели — 0.5 балла**  
   - Выберите энкодерную модель для дообучения  

In [67]:
flush()  # Очищаем память GPU перед загрузкой модели

# Загружаем предобученную мультилингвальную модель SentenceTransformer
base_name = "intfloat/multilingual-e5-small"
st_model = SentenceTransformer(base_name, device=device)

# Извлекаем базовую модель (AutoModel) из SentenceTransformer
backbone = st_model[0].auto_model

# Включаем gradient checkpointing для экономии памяти при обучении
if hasattr(backbone, "gradient_checkpointing_enable"):
    backbone.gradient_checkpointing_enable()

# Настраиваем параметры LoRA (Low-Rank Adaptation) для дообучения модели
lora_cfg = LoraConfig(
    r=16,                        # ранг матриц LoRA
    lora_alpha=16,                # масштабирование LoRA весов
    lora_dropout=0.05,            # dropout для LoRA слоёв
    bias="none",                  # тип обучения bias (здесь не обучаем)
    target_modules=["query", "key", "value", "dense"],  # какие модули адаптировать
    task_type=TaskType.FEATURE_EXTRACTION,             # тип задачи (извлечение признаков)
)

# Применяем LoRA к backbone модели
peft_backbone = get_peft_model(backbone, lora_cfg)

# Выводим количество обучаемых параметров после применения LoRA
peft_backbone.print_trainable_parameters()

# Подменяем оригинальный backbone на PEFT-модель
st_model[0].auto_model = peft_backbone

# Определяем функцию потерь MultipleNegativesRankingLoss для контрастивного обучения
loss_fn = losses.MultipleNegativesRankingLoss(st_model)


modules.json:   0%|          | 0.00/387 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/655 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/471M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/443 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/167 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/200 [00:00<?, ?B/s]

trainable params: 1,339,392 || all params: 118,993,152 || trainable%: 1.1256


3. **Предварительная оценка качества — 1 балл**  
   - Выделите валидационное множество, проверьте на нем качество поиска с помощью метрик @k  
   - Также для валидации можно использовать down-stream задачу на эмбеддингах из модели, если она у вас есть  

In [68]:
# Функция для получения эмбеддингов (векторных представлений) текстов
def embed(texts, model, batch_size=128, normalize=True):
    vectors = model.encode(
        texts,
        batch_size=batch_size,           # количество примеров в одном батче
        convert_to_numpy=True,           # возвращаем результат в формате numpy
        normalize_embeddings=normalize,  # нормализация векторов (нужно для косинусного сходства)
        device=device,                   # устройство (CPU/GPU)
        show_progress_bar=False,         # не показывать прогресс-бар
    )
    return vectors

# Разделяем пары (query, document) на отдельные списки
queries = [q for q, _ in val_pairs]   # список запросов
docs    = [d for _, d in val_pairs]   # список документов (ответов)

# Строим эмбеддинги для запросов и документов
q_vecs = embed(queries, st_model)     # матрица эмбеддингов запросов
d_vecs = embed(docs, st_model)        # матрица эмбеддингов документов

# Считаем матрицу сходств между запросами и документами
# Размер будет (число_запросов, число_документов)
sims = np.matmul(q_vecs, d_vecs.T)

# Берём топ-5 самых похожих документов для каждого запроса
k = min(5, sims.shape[1])  # ограничиваем k (не больше числа документов)
topk_idx = np.argpartition(-sims, kth=k-1, axis=1)[:, :k]

# Индексы правильных документов (правильный ответ для запроса i — это документ i)
true_idx = np.arange(len(val_pairs))

# Проверяем, есть ли правильный документ среди топ-5 кандидатов
hits = (topk_idx == true_idx[:, None]).any(axis=1)

# Доля правильных ответов среди всех запросов = метрика Hit@5
hit5 = hits.mean()
print(f"Hit@5: {hit5:.3f}")


Hit@5: 0.716


In [69]:
train_data = [
    {
        "anchor": q,   # "anchor" — это вопрос или входная инструкция
        "positive": d  # "positive" — это правильный ответ или контекст для якоря
    }
    for q, d in train_pairs  # проходим по всем парам (вопрос, ответ) из тренировочного набора
]

# Создаём объект Dataset из списка словарей
train_ds = Dataset.from_list(train_data)


val_data = [
    {
        "anchor": q,   # аналогично для валидационного набора
        "positive": d
    }
    for q, d in val_pairs
]
val_ds = Dataset.from_list(val_data)


4. **QLora-дообучение — 2 балла**  
   - Настройте параметры дообучения через QLora, опишите свой выбор значений и настраиваемых параметров в комментариях  
   - Обучите модель на вашем датасете  

## Параметры дообучения модели (QLoRA)

| Параметр                | Значение                 | Обоснование                                                                                             |
| ----------------------- | ------------------------ | ------------------------------------------------------------------------------------------------------- |
| **LoRA `r`**            | 16                       | Баланс между качеством и потреблением памяти (меньше — хуже качество, больше — выше VRAM).              |
| **LoRA α**              | 16                       | Масштабирование весов адаптации, равное `r`, чтобы не усиливать переобучение.                           |
| **LoRA dropout**        | 0.05                     | Небольшой регуляризатор, предотвращает переобучение на парах.                                           |
| **bias**                | none                     | Не обучаем bias — это экономит память и почти не влияет на результат.                                   |
| **target\_modules**     | query, key, value, dense | Основные слои внимания и линейные проекции — дают наибольший прирост качества при минимальных затратах. |
| **epochs**              | 5                        | Достаточно для доадаптации предобученной модели, больше может привести к переобучению.                  |
| **batch\_size**         | 32                       | Умеренное значение: не перегружает GPU, сохраняет стабильность обучения.                                |
| **grad. accumulation**  | 4                        | Эффективный батч = 128, улучшает стабильность градиентов без лишней VRAM.                               |
| **max\_steps\_cap**     | 120                      | Ограничение шагов для ускорения экспериментов и контроля переобучения.                                  |
| **warmup\_ratio**       | 0.05                     | Постепенный разгон LR на 5% шагов — предотвращает резкие скачки на старте.                              |
| **learning rate**       | 2e-4                     | Относительно высокий LR, т.к. обучаются только LoRA-слои (а не вся модель).                             |
| **scheduler**           | cosine                   | Плавное уменьшение LR, хорошо подходит для дообучения.                                                  |
| **weight\_decay**       | 0.01                     | Лёгкая регуляризация, предотвращает переобучение.                                                       |
| **optim**               | paged\_adamw\_8bit       | Экономит VRAM и ускоряет обучение за счёт 8-битной квантизации.                                         |
| **fp16**                | True (если GPU)          | Снижает нагрузку на память и ускоряет вычисления.                                                       |
| **grad. checkpointing** | True                     | Экономия памяти (чуть медленнее обучение).                                                              |
| **logging\_steps**      | 10                       | Частое логирование для мониторинга.                                                                     |
| **eval\_steps**         | 50                       | Промежуточная оценка динамики качества.                                                                 |
| **seed**                | 42                       | Фиксируем для воспроизводимости.                                                                        |


## Обоснование выбора параметров QLoRA-обучения

Для дообучения энкодерной модели использовался метод **QLoRA** совместно с адаптерами **LoRA**, что позволяет эффективно обучать большие модели для извлечения эмбеддингов с минимальными затратами памяти. Ниже приведено обоснование ключевых параметров:

1. **LoRA-адаптеры**

   * `r=16` и `lora_alpha=16` — стандартные значения, обеспечивающие баланс между качеством обучения и количеством обучаемых параметров.
   * `lora_dropout=0.05` — лёгкая регуляризация для предотвращения переобучения на ограниченном контрастивном датасете.
   * В `target_modules` выбраны критически важные слои (`query`, `key`, `value`, `dense`), так как именно они формируют качественные эмбеддинги текстов.
   * `bias="none"` — смещение не обучается, чтобы снизить использование GPU-памяти.
   * `task_type=TaskType.FEATURE_EXTRACTION` — задаёт контрастивную задачу извлечения признаков, подходящую для обучения на парах (anchor+positive).

2. **Оптимизация памяти**

   * Включён **gradient checkpointing** (`gradient_checkpointing=True`), что позволяет экономить видеопамять при обучении длинных последовательностей.
   * Использован 8-битный оптимизатор `paged_adamw_8bit`, который снижает потребление памяти без потери точности градиентов.
   * FP16 (`fp16=True`) включён для ускорения вычислений и уменьшения объёма видеопамяти.

3. **Гиперпараметры обучения**

   * `epochs=5` — достаточно для адаптации модели на специализированном датасете без переобучения.
   * `batch_size=32` и `gradient_accumulation_steps=4` — реальный размер батча ограничен памятью, но аккумуляция градиентов позволяет обучать модель как при батче 128.
   * `learning_rate=2e-4` — умеренно высокий шаг обучения для LoRA-адаптеров, чтобы быстро адаптировать эмбеддинги.
   * `warmup_ratio=0.05` — плавный разогрев learning rate для предотвращения резких скачков градиентов в начале.
   * `weight_decay=0.01` — лёгкая регуляризация для улучшения обобщающей способности.
   * `lr_scheduler_type="cosine"` — косинусный планировщик, позволяющий постепенно уменьшать LR по мере обучения.
   * `max_steps=120` на эпоху — ограничение числа шагов для стабильного и контролируемого обучения.

4. **Организация эксперимента**

   * `logging_steps=10` — логирование каждые 10 шагов для мониторинга обучения.
   * `eval_steps=50` — промежуточная оценка качества на валидационном датасете.
   * `output_dir="st-encoder-qlora-out"` — сохранение модели и логов в отдельную директорию.
   * `report_to="none"` — отключена интеграция с внешними сервисами логирования.
   * `seed=42` — фиксированное значение генератора случайных чисел для воспроизводимости.

---

Таким образом, выбранные параметры позволяют провести **ресурсно-эффективное дообучение энкодерной модели**: минимальное использование GPU-памяти, стабильность обучения и улучшение качества эмбеддингов для downstream-задач.


In [70]:
# -------------------------
# Настройка гиперпараметров
# -------------------------
epochs = 5                       # Количество эпох обучения
batch_size = 32                   # Размер батча
gradient_accumulation_steps = 4   # Количество шагов накопления градиента
max_steps_cap = 120               # Максимальное количество шагов на эпоху
warmup_ratio = 0.05               # Доля шагов для разогрева learning rate

# -------------------------
# Вычисляем количество шагов
# -------------------------
steps_per_epoch = min(math.ceil(len(train_ds) / batch_size), max_steps_cap)
total_steps = steps_per_epoch * epochs  # Общее количество шагов обучения

# -------------------------
# Определяем функцию потерь
# -------------------------
# MultipleNegativesRankingLoss — контрастивная функция потерь
# Использует негативные примеры из других элементов батча
loss_fn = losses.MultipleNegativesRankingLoss(st_model)

# -------------------------
# Настройки обучения модели
# -------------------------
training_args = SentenceTransformerTrainingArguments(
    output_dir="st-encoder-qlora-out",   # Папка для сохранения модели
    per_device_train_batch_size=batch_size,  # Размер батча на устройство
    gradient_accumulation_steps=gradient_accumulation_steps,  # Накопление градиентов
    learning_rate=2e-4,                 # Начальная скорость обучения
    warmup_ratio=warmup_ratio,           # Разогрев learning rate
    num_train_epochs=epochs,             # Количество эпох
    max_steps=total_steps,               # Максимальное число шагов
    lr_scheduler_type="cosine",          # Косинусный scheduler для LR
    weight_decay=0.01,                   # Регуляризация весов
    logging_steps=10,                     # Логирование каждые 10 шагов
    save_strategy="no",                   # Не сохраняем промежуточные чекпоинты
    eval_strategy="steps",                # Оценка модели по шагам
    eval_steps=50,                        # Оценка каждые 50 шагов
    report_to="none",                     # Не отправлять логи в внешние сервисы
    optim="paged_adamw_8bit",            # Оптимизатор с 8-bit AdamW для экономии памяти
    fp16=torch.cuda.is_available(),       # Используем FP16 если есть GPU
    gradient_checkpointing=True,          # Экономия памяти с помощью gradient checkpointing
    dataloader_drop_last=True,            # Отбрасываем неполные батчи
    dataloader_num_workers=0,             # Количество потоков для DataLoader
    seed=42,                              # Фиксируем seed для воспроизводимости
)

# -------------------------
# Создаем тренер
# -------------------------
trainer = SentenceTransformerTrainer(
    model=st_model,            # Модель для обучения
    args=training_args,        # Аргументы обучения
    train_dataset=train_ds,    # Датасет для тренировки
    eval_dataset=val_ds,       # Валидационный датасет
    loss=loss_fn,              # Функция потерь
)

# -------------------------
# Запуск обучения
# -------------------------
trainer.train()

# -------------------------
# Сохраняем финальную модель
# -------------------------
st_model.save("st-encoder-qlora-out/final_model")


Computing widget examples:   0%|          | 0/1 [00:00<?, ?example/s]

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 2,000 | Num Epochs = 20 | Total steps = 315
O^O/ \_/ \    Batch size per device = 32 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (32 x 4 x 1) = 128
 "-____-"     Trainable parameters = 1,339,392 of 118,993,152 (1.13% trained)


5. **Оценка качества обучения — 1 балл**  
   - Проверьте качество генерации на валидационном датасете 

In [71]:
# 1. Получаем эмбеддинги для запросов и документов после обучения
q_vecs_after = embed(queries, st_model)  # Векторы для всех запросов
d_vecs_after = embed(docs, st_model)     # Векторы для всех документов

# 2. Считаем схожесть между каждым запросом и каждым документом
# np.matmul выполняет матричное умножение: q_vecs * d_vecs^T
sims_after = np.matmul(q_vecs_after, d_vecs_after.T)

# 3. Находим индексы топ-k наиболее похожих документов для каждого запроса
k = min(5, sims_after.shape[1])  # Берем максимум 5
topk_idx = np.argpartition(-sims_after, kth=k-1, axis=1)[:, :k]

# 4. Индексы "правильных" документов (предполагается, что они по порядку)
true_idx = np.arange(len(val_pairs))

# 5. Проверяем, попал ли правильный документ в топ-k
hits = (topk_idx == true_idx[:, None]).any(axis=1)

# 6. Среднее по всем запросам — метрика Hit@5
hit5 = hits.mean()

# 7. Выводим результат
print(f"Hit@5: {hit5:.3f}")


Hit@5: 0.782


# Отчёт

### 1. Выбор датасета

Для обучения была выбрана задача контрастивного обучения, где есть пары *запрос–документ*.
Использован датасет `akoksal/muri-it-language-split`, содержащий пары (query, positive\_document).
Для валидации применялись метрики поиска (Hit\@k).

---

### 2. Выбор предобученной модели

В качестве базовой модели был выбран **`intfloat/multilingual-e5-small`** — многоязычный энкодер из семейства **E5**, хорошо зарекомендовавший себя в задачах семантического поиска.
Причины выбора:

* компактный размер, подходящий для экспериментов;
* поддержка русского языка;
* хорошее качество базовых эмбеддингов.

---

### 3. Предварительная оценка качества

Для оценки качества использовалась метрика **Hit\@5** (наличие правильного документа среди топ-5 ближайших по косинусному сходству).

* **До обучения:** `Hit@5 = 0.716`

Это базовая точка отсчёта.

---

### 4. QLoRA-дообучение

Для дообучения применялся подход **PEFT с LoRA**, что позволяет обучать только малую часть параметров, сохраняя эффективность модели.

#### Настройка параметров LoRA:

* **r=16** — размер низкоранговых матриц (компромисс между качеством и памятью);
* **lora\_alpha=16** — коэффициент масштабирования LoRA;
* **lora\_dropout=0.05** — регуляризация для повышения устойчивости;
* **bias="none"** — bias не дообучаем;
* **target\_modules=\["query", "key", "value", "dense"]** — адаптируем ключевые модули трансформера;
* **task\_type=FEATURE\_EXTRACTION** — задача на извлечение эмбеддингов.

#### Функция потерь:

Использовалась **MultipleNegativesRankingLoss**, стандартная для контрастивного обучения. Она минимизирует расстояние между query и его положительным документом, одновременно отталкивая от других документов в батче.

---

### 5. Оценка качества обучения

После дообучения повторно проведена проверка на валидационном множестве.

* **После обучения:** `Hit@5 = 0.782`

Таким образом, удалось повысить качество поиска на \~6.6 процентных пунктов.

---

### Вывод

Дообучение энкодерной модели **`multilingual-e5-small`** с помощью **QLoRA** и контрастивного обучения дало заметное улучшение качества на downstream-задаче семантического поиска:

* модель лучше находит релевантные документы для запросов;
* метрика Hit\@5 выросла с **0.716 → 0.782**;
* использование LoRA позволило провести обучение эффективно и без значительного роста потребления GPU-памяти.
